<a href="https://colab.research.google.com/github/bish-ai/Bishal.py/blob/main/heart_disease_predictor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [29]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
cancer_data=pd.read_csv("https://storage.googleapis.com/kagglesdsdata/datasets/888463/1508992/heart_disease_uci.csv?X-Goog-Algorithm=GOOG4-RSA-SHA256&X-Goog-Credential=gcp-kaggle-com%40kaggle-161607.iam.gserviceaccount.com%2F20260123%2Fauto%2Fstorage%2Fgoog4_request&X-Goog-Date=20260123T111824Z&X-Goog-Expires=259200&X-Goog-SignedHeaders=host&X-Goog-Signature=377971cb0546ad7b192b994052ba49002a726b984b982ef1261054a613bb5012d8b4a01569e22671a8317de8eb28c4eab6703521cfeaf6b57e9bf6ab1fb37c5bba0314187bcc86b1fb74dd0d19d648e612b463a432937eaa33ff1a0d3a46bc939d9ba1ef3f90475777513c6a828d7a41071b03457d28a2c31b3e30f6d601103850101dfcfa710f3211d2445c782226638536381c178e2c5deb8304debc8607e8c1ca145d8c3c157eb92874f1b362df19c2a3d64a3332b06a522ebbf76a434fabc303452009fc88baeb366c15d79c3cc7ac25fce94174f1b621c9c3d433052682c3ee400512ada47b3efc0391f78ccee87d834468fff75d6f96dca65fb1bcce99")
cancer_data.tail()
cancer_data.head()
cancer_data.shape
cancer_data.info()
cancer_data.duplicated()
cancer_data.isnull().sum()
cancer_data.head()
cancer_data["num"].value_counts()
cancer_data.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 920 entries, 0 to 919
Data columns (total 16 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   id        920 non-null    int64  
 1   age       920 non-null    int64  
 2   sex       920 non-null    object 
 3   dataset   920 non-null    object 
 4   cp        920 non-null    object 
 5   trestbps  861 non-null    float64
 6   chol      890 non-null    float64
 7   fbs       830 non-null    object 
 8   restecg   918 non-null    object 
 9   thalch    865 non-null    float64
 10  exang     865 non-null    object 
 11  oldpeak   858 non-null    float64
 12  slope     611 non-null    object 
 13  ca        309 non-null    float64
 14  thal      434 non-null    object 
 15  num       920 non-null    int64  
dtypes: float64(5), int64(3), object(8)
memory usage: 115.1+ KB


id  age     sex    dataset               cp  trestbps   chol    fbs  \
0   1   63    Male  Cleveland   typical angina     145.0  233.0   True   
1   2   67    Male  Cleveland     asymptomatic     160.0  286.0  False   
2   3   67    Male  Cleveland     asymptomatic     120.0  229.0  False   
3   4   37    Male  Cleveland      non-anginal     130.0  250.0  False   
4   5   41  Female  Cleveland  atypical angina     130.0  204.0  False   

          restecg  thalch  exang  oldpeak        slope   ca  \
0  lv hypertrophy   150.0  False      2.3  downsloping  0.0   
1  lv hypertrophy   108.0   True      1.5         flat  3.0   
2  lv hypertrophy   129.0   True      2.6         flat  2.0   
3          normal   187.0  False      3.5  downsloping  0.0   
4  lv hypertrophy   172.0  False      1.4    upsloping  0.0   

                thal  num  
0       fixed defect    0  
1             normal    2  
2  reversable defect    1  
3             normal    0  
4             normal    0

In [82]:
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder

# split
x = cancer_data.drop(columns=["num"])
y = cancer_data[["num"]]

x_train, x_test, y_train, y_test = train_test_split(
    x, y, train_size=0.7, random_state=43
)

# Identify numerical and categorical columns from the entire dataset
# This ensures we catch all columns that might need imputation before splitting
# or that need imputation for the categorical portion.
all_numerical_cols = x.select_dtypes(include=["int64", "float64"]).columns
all_categorical_cols = x.select_dtypes(include=["object"]).columns

# Impute numerical columns using SimpleImputer
si_numerical = SimpleImputer(strategy="mean")
x_train[all_numerical_cols] = si_numerical.fit_transform(x_train[all_numerical_cols])
x_test[all_numerical_cols] = si_numerical.transform(x_test[all_numerical_cols])

# Impute categorical columns using SimpleImputer (most_frequent strategy)
si_categorical = SimpleImputer(strategy='most_frequent')
x_train[all_categorical_cols] = si_categorical.fit_transform(x_train[all_categorical_cols])
x_test[all_categorical_cols] = si_categorical.transform(x_test[all_categorical_cols])

# ColumnTransformer for encoding categorical features
ct = ColumnTransformer(
    transformers=[
        ("sex", OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1), ["sex"]),
        ("dataset", OneHotEncoder(handle_unknown="ignore"), ["dataset"]),
        ("cp", OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1), ["cp"]),
        ("fbs", OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1), ["fbs"]),
        ("restecg", OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1), ["restecg"]),
        ("exang", OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1), ["exang"]),
        ("slope", OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1), ["slope"]),
        ("thal", OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1), ["thal"]),
    ],
    remainder="passthrough"
)

x_train_transformed = ct.fit_transform(x_train)
x_test_transformed = ct.transform(x_test)

from sklearn.preprocessing import StandardScaler
ss = StandardScaler()
x_train_scaled = ss.fit_transform(x_train_transformed)
x_test_scaled = ss.transform(x_test_transformed)

In [86]:
from sklearn.linear_model import LogisticRegression
lr = LogisticRegression()
y_train_pd = pd.DataFrame(y_train)
y_test_pd = pd.DataFrame(y_test)
y_train = y_train_pd.values.ravel()
y_test = y_test_pd.values.ravel()
lr_x_y_train_fit = lr.fit(x_train_scaled, y_train)
lr_x_y_train_predict= lr.predict(x_train_scaled)
from sklearn.metrics import r2_score
r2_score(y_train,lr_x_y_train_predict)

0.3208946541257991

# Task
Create a Logistic Regression model to classify heart disease based on the provided dataset. Begin by separating the features and target variable, then split the data into training and testing sets. Preprocess the data by identifying numerical and categorical columns. For numerical features, use `SimpleImputer` to handle missing values. For categorical features, apply `SimpleImputer` with a 'most_frequent' strategy, then use `OneHotEncoder` for nominal features (`sex`, `dataset`) and `OrdinalEncoder` for ordinal features (`cp`, `fbs`, `restecg`, `exang`, `slope`, `thal`), ensuring `restecg` is correctly spelled. Finally, train the `LogisticRegression` model on the preprocessed training data and evaluate its performance using accuracy score on the preprocessed test data.

## Identify Categorical and Numerical Columns

### Subtask:
Separate the columns in `x_train` and `x_test` into numerical and categorical types for appropriate preprocessing.


**Reasoning**:
I will define the `numerical_cols`, `nominal_categorical_cols`, and `ordinal_categorical_cols` lists as specified in the instructions to categorize the columns for preprocessing.



In [37]:
numerical_cols = ['age', 'trestbps', 'chol', 'thalch', 'oldpeak', 'ca']
nominal_categorical_cols = ['sex', 'dataset']
ordinal_categorical_cols = ['cp', 'fbs', 'restecg', 'exang', 'slope', 'thal']

print("Numerical columns identified:", numerical_cols)
print("Nominal categorical columns identified:", nominal_categorical_cols)
print("Ordinal categorical columns identified:", ordinal_categorical_cols)

Numerical columns identified: ['age', 'trestbps', 'chol', 'thalch', 'oldpeak', 'ca']
Nominal categorical columns identified: ['sex', 'dataset']
Ordinal categorical columns identified: ['cp', 'fbs', 'restecg', 'exang', 'slope', 'thal']


## Define Preprocessing Pipelines

### Subtask:
Create a ColumnTransformer that handles both numerical and categorical features. For numerical features, use SimpleImputer to fill missing values. For categorical features, use SimpleImputer with a 'most_frequent' strategy, then OneHotEncoder for nominal features and OrdinalEncoder for ordinal features, ensuring 'restecg' is correctly spelled. Apply the transformer to `x_train` and `x_test`.


**Reasoning**:
The previous code failed because the ColumnTransformer was applied to `y_train` instead of `x_train`, and `LabelEncoder` was used directly in the `ColumnTransformer` which is not suitable for this context. I need to correctly define preprocessing pipelines for numerical, nominal, and ordinal categorical features using `SimpleImputer`, `OneHotEncoder`, and `OrdinalEncoder` within a `ColumnTransformer` and then apply it to `x_train` and `x_test`.



In [39]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Create preprocessing pipelines
numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean'))
])

nominal_categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
])

ordinal_categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal', OrdinalEncoder(handle_unknown='ignore'))
])

# Create a ColumnTransformer to apply different transformations to different columns
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_cols),
        ('nom_cat', nominal_categorical_transformer, nominal_categorical_cols),
        ('ord_cat', ordinal_categorical_transformer, ordinal_categorical_cols)
    ],
    remainder='passthrough' # Keep other columns not specified (like 'id' if not dropped)
)

# Apply the ColumnTransformer to x_train and x_test
x_train_processed = preprocessor.fit_transform(x_train)
x_test_processed = preprocessor.transform(x_test)

print("Shape of processed x_train:", x_train_processed.shape)
print("Shape of processed x_test:", x_test_processed.shape)

InvalidParameterError: The 'handle_unknown' parameter of OrdinalEncoder must be a str among {'error', 'use_encoded_value'}. Got 'ignore' instead.

**Reasoning**:
The previous code failed because `handle_unknown='ignore'` is not a valid parameter for `OrdinalEncoder`. The error message specifies that it must be `'error'` or `'use_encoded_value'`. I will change it to `'use_encoded_value'` to handle unseen categories during transformation without raising an error.



In [40]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# Create preprocessing pipelines
numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean'))
])

nominal_categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
])

ordinal_categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal', OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1))
])

# Create a ColumnTransformer to apply different transformations to different columns
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_cols),
        ('nom_cat', nominal_categorical_transformer, nominal_categorical_cols),
        ('ord_cat', ordinal_categorical_transformer, ordinal_categorical_cols)
    ],
    remainder='passthrough' # Keep other columns not specified (like 'id' if not dropped)
)

# Apply the ColumnTransformer to x_train and x_test
x_train_processed = preprocessor.fit_transform(x_train)
x_test_processed = preprocessor.transform(x_test)

print("Shape of processed x_train:", x_train_processed.shape)
print("Shape of processed x_test:", x_test_processed.shape)

Shape of processed x_train: (644, 19)
Shape of processed x_test: (276, 19)


**Reasoning**:
The preprocessing of `x_train` and `x_test` is complete. Now, I will train a `LogisticRegression` model on the processed training data and evaluate its performance using accuracy score on the processed test data, as per the main task.



In [41]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Initialize and train the Logistic Regression model
model = LogisticRegression(max_iter=1000) # Increased max_iter for convergence
model.fit(x_train_processed, y_train)

# Make predictions on the processed test data
y_pred = model.predict(x_test_processed)

# Evaluate the model's accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Model accuracy on the test set: {accuracy:.4f}")

/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:1408: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Model accuracy on the test set: 0.5688


/usr/local/lib/python3.12/dist-packages/sklearn/linear_model/_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


**Reasoning**:
The previous Logistic Regression model raised a `ConvergenceWarning`, indicating that the optimization algorithm did not converge within the specified number of iterations. The warning suggests scaling the data to help with convergence. I will modify the numerical preprocessing pipeline to include `StandardScaler` to scale the numerical features. Additionally, I will address the `DataConversionWarning` by flattening `y_train` using `.values.ravel()` before fitting the model.



In [42]:
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

# Create preprocessing pipelines
numerical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='mean')),
    ('scaler', StandardScaler()) # Add StandardScaler for numerical features
])

nominal_categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore', sparse_output=False))
])

ordinal_categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('ordinal', OrdinalEncoder(handle_unknown='use_encoded_value', unknown_value=-1))
])

# Create a ColumnTransformer to apply different transformations to different columns
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numerical_transformer, numerical_cols),
        ('nom_cat', nominal_categorical_transformer, nominal_categorical_cols),
        ('ord_cat', ordinal_categorical_transformer, ordinal_categorical_cols)
    ],
    remainder='passthrough' # Keep other columns not specified (like 'id' if not dropped)
)

# Apply the ColumnTransformer to x_train and x_test
x_train_processed = preprocessor.fit_transform(x_train)
x_test_processed = preprocessor.transform(x_test)

# Initialize and train the Logistic Regression model
# Increased max_iter for convergence, and will now scale data
model = LogisticRegression(max_iter=5000, solver='liblinear') # Using a different solver that might converge better with scaled data, and further increased max_iter
model.fit(x_train_processed, y_train.values.ravel()) # Flatten y_train to suppress DataConversionWarning

# Make predictions on the processed test data
y_pred = model.predict(x_test_processed)

# Evaluate the model's accuracy
accuracy = accuracy_score(y_test, y_pred)
print(f"Model accuracy on the test set: {accuracy:.4f}")

Model accuracy on the test set: 0.5543


# Task
Summarize the simplified data preprocessing steps, including imputation and scaling for numerical features, and imputation and encoding for categorical features, along with the final accuracy of the Logistic Regression model.

## Impute Numerical Features

### Subtask:
Apply `SimpleImputer` with a 'mean' strategy to the numerical columns in both `x_train` and `x_test`.


# Task
Summarize the simplified data preprocessing steps, including imputation and scaling for numerical features (age, trestbps, chol, thalch, oldpeak, ca), and imputation and encoding for categorical features (nominal: sex, dataset; ordinal: cp, fbs, restecg, exang, slope, thal), along with the final accuracy of the Logistic Regression model.

## Impute Numerical Features

### Subtask:
Describe the imputation strategy used for numerical features in the preprocessing steps.


### Imputation Strategy for Numerical Features

Missing values in the numerical columns (`age`, `trestbps`, `chol`, `thalch`, `oldpeak`, `ca`) were imputed using `SimpleImputer` with a 'mean' strategy. This approach replaces any missing data points within these columns with the mean value of their respective columns, providing a central tendency estimate for the missing data.

## Final Task

### Subtask:
Summarize the simplified data preprocessing steps, highlighting each transformation technique used, and report the final accuracy of the Logistic Regression model.


## Summary:

### Data Analysis Key Findings

*   Missing values in numerical features, specifically `age`, `trestbps`, `chol`, `thalch`, `oldpeak`, and `ca`, were addressed.
*   The imputation strategy employed was `SimpleImputer` with a 'mean' strategy, where missing data points in these columns were replaced by the mean value of their respective columns.

### Insights or Next Steps

*   The chosen imputation method provides a basic way to handle missing numerical data by using central tendency, which might be further refined with more sophisticated techniques depending on data distribution and impact on model performance.


# Task
Summarize the simplified data preprocessing steps, including imputation and scaling for numerical features (age, trestbps, chol, thalch, oldpeak, ca), and imputation and encoding for categorical features (nominal: sex, dataset; ordinal: cp, fbs, restecg, exang, slope, thal), along with the final accuracy of the Logistic Regression model.

## Summarize the simplified data preprocessing steps, highlighting each transformation technique used, and report the final accuracy of the Logistic Regression model.

### Subtask:
Summarize the data preprocessing steps and report the final accuracy of the Logistic Regression model.


## Summary of Data Preprocessing and Model Performance

The data preprocessing involved a `ColumnTransformer` to apply different transformations to numerical and categorical features:

### 1. Numerical Feature Preprocessing:
*   **Imputation**: Missing values in numerical columns (`age`, `trestbps`, `chol`, `thalch`, `oldpeak`, `ca`) were handled using `SimpleImputer` with a 'mean' strategy.
*   **Scaling**: After imputation, numerical features were scaled using `StandardScaler` to standardize their range, which aids in the convergence of many machine learning algorithms like Logistic Regression.

### 2. Categorical Feature Preprocessing:
*   **Imputation**: For all categorical features (both nominal and ordinal), missing values were imputed using `SimpleImputer` with a 'most_frequent' strategy.
*   **Nominal Features**: For nominal categorical columns (`sex`, `dataset`), `OneHotEncoder` was applied with `handle_unknown='ignore'` to convert them into a binary vector representation.
*   **Ordinal Features**: For ordinal categorical columns (`cp`, `fbs`, `restecg`, `exang`, `slope`, `thal`), `OrdinalEncoder` was used with `handle_unknown='use_encoded_value'` and `unknown_value=-1` to assign numerical ranks.

### 3. Model Training:
*   A `LogisticRegression` model was initialized with `max_iter=5000` and `solver='liblinear'`. The model was trained on the preprocessed `x_train` data and `y_train` (flattened using `.values.ravel()` to address a `DataConversionWarning`).

### Final Model Accuracy:
*   The Logistic Regression model achieved an accuracy of **0.5543** on the preprocessed test set.

## Final Task

### Subtask:
The task has been completed. The summary of data preprocessing steps and the final accuracy of the Logistic Regression model have been provided in the previous turn.


## Summary:

### Data Analysis Key Findings

*   **Numerical Feature Preprocessing**: Missing values in numerical columns (`age`, `trestbps`, `chol`, `thalch`, `oldpeak`, `ca`) were imputed using the 'mean' strategy, followed by scaling using `StandardScaler`.
*   **Categorical Feature Preprocessing**:
    *   Missing values in all categorical features were imputed using the 'most\_frequent' strategy.
    *   Nominal features (`sex`, `dataset`) were encoded using `OneHotEncoder`.
    *   Ordinal features (`cp`, `fbs`, `restecg`, `exang`, `slope`, `thal`) were encoded using `OrdinalEncoder`.
*   **Model Performance**: The Logistic Regression model achieved a final accuracy of 0.5543 on the preprocessed test set.

### Insights or Next Steps

*   The current Logistic Regression model's accuracy of 55.43% suggests that its predictive power is only slightly better than random chance (if classes are balanced) for the given task.
*   Further steps should include exploring other machine learning algorithms, performing hyperparameter tuning for the Logistic Regression model, or investigating more advanced feature engineering techniques to improve performance.
